src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,165 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int count;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail != null) {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        if (head == null) {
+            head = newNode;
+        }
+        count++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexToAdd(index);
+        if (index == size()) {
+            add(value);
+        } else if (index == 0) {
+            addToHead(value);
+        } else {
+            Node<T> source = getNodeByIndex(index);
+            Node<T> toAdd = new Node<>(source.previous, value, source);
+            source.previous.next = toAdd;
+            source.previous = toAdd;
+            count++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexToGet(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexToGet(index);
+        Node<T> toSet = getNodeByIndex(index);
+        T oldValue = toSet.value;
+        toSet.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexToGet(index);
+        Node<T> toRemove = getNodeByIndex(index);
+        unlink(toRemove);
+        return toRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        return unlink(getNodeByValue(object));
     }
 
     @Override
     public int size() {
-        return 0;
+        return count;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    private void checkIndexToGet(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Invalid index " + index + " for size " + size());
+        }
+    }
+
+    private void checkIndexToAdd(int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException(
+                    String.format("Index %d out of bound for size %d", index, size()));
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        return index <= size() / 2 ? getNodeFromHead(index) :
+                getNodeFromTail(index);
+    }
```

Comment:
I would simplify the name of this method to just `node(int index)`

and also use proper formatting for ternary operator

take a look, doesn't it look better?)
```suggestion
    private Node<T> node(int index) {
        return index <= size / 2 ? nodeFromHead(index) : nodeFromTail(index);
    }
```